From 6f6ef47a526e09e09631234b5f39f06340b8c879 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Mon, 17 Apr 2017 19:02:50 +0800
Subject: [PATCH 1706/5242] MLK-14703-3 video: mipi_dsi_northwest: refine some
 configs via traffic mode.

commit  8ebe51fe72e2f3f00005e7e949d348e6242db58a from
https://source.codeaurora.org/external/imx/linux-imx.git

Refine some DSI registers configuration by using the traffic
mode:
   . set 'DSI_VIDEO_MODE' directly from 'traffic_mode'.
   . set 'HFP', 'HBP' and 'HSA' by distinguishing traffic modes.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit 2549b6f527c462d0a556dceb36c4c3ce7470e809)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/mipi_dsi.h           |    4 ++++
 drivers/video/fbdev/mxc/mipi_dsi_northwest.c |   30 +++++++++++++++++---------
 2 files changed, 24 insertions(+), 10 deletions(-)

diff --git a/drivers/video/fbdev/mxc/mipi_dsi.h b/drivers/video/fbdev/mxc/mipi_dsi.h
index 04008b3..6cbdd70 100644
--- a/drivers/video/fbdev/mxc/mipi_dsi.h
+++ b/drivers/video/fbdev/mxc/mipi_dsi.h
@@ -25,6 +25,10 @@
 
 #define	DSI_CMD_BUF_MAXSIZE         (128)
 
+#define DSI_NON_BURST_WITH_SYNC_PULSE  0
+#define DSI_NON_BURST_WITH_SYNC_EVENT  1
+#define DSI_BURST_MODE                 2
+
 /* DPI interface pixel color coding map */
 enum mipi_dsi_dpi_fmt {
 	MIPI_RGB565_PACKED = 0,
diff --git a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
index d710a97..904823b 100644
--- a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
+++ b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
@@ -417,18 +417,28 @@ static int mipi_dsi_dpi_init(struct mipi_dsi_info *mipi_dsi)
 	writel(pixel_fmt, mipi_dsi->mmio_base + DPI_PIXEL_FORMAT);
 	writel(0x0, mipi_dsi->mmio_base + DPI_VSYNC_POLARITY);
 	writel(0x0, mipi_dsi->mmio_base + DPI_HSYNC_POLARITY);
-	writel(mipi_dsi->encoder ? 0x0 : 0x2,
+	writel(mipi_dsi->traffic_mode,
 	       mipi_dsi->mmio_base + DPI_VIDEO_MODE);
 
-	writel(mipi_dsi->encoder ?
-	       0x10 : mode->right_margin * (bpp >> 3),
-	       mipi_dsi->mmio_base + DPI_HFP);
-	writel(mipi_dsi->encoder ?
-	       0x60 : mode->left_margin * (bpp >> 3),
-	       mipi_dsi->mmio_base + DPI_HBP);
-	writel(mipi_dsi->encoder ?
-	       0xf0 : mode->hsync_len * (bpp >> 3),
-	       mipi_dsi->mmio_base + DPI_HSA);
+	switch (mipi_dsi->traffic_mode) {
+	case DSI_NON_BURST_WITH_SYNC_PULSE:
+		writel(0x10, mipi_dsi->mmio_base + DPI_HFP);
+		writel(0x60, mipi_dsi->mmio_base + DPI_HBP);
+		writel(0xf0, mipi_dsi->mmio_base + DPI_HSA);
+		break;
+	case DSI_BURST_MODE:
+		writel(mode->right_margin * (bpp >> 3),
+		       mipi_dsi->mmio_base + DPI_HFP);
+		writel(mode->left_margin * (bpp >> 3),
+		       mipi_dsi->mmio_base + DPI_HBP);
+		writel(mode->hsync_len * (bpp >> 3),
+		       mipi_dsi->mmio_base + DPI_HSA);
+		break;
+	default:
+		pr_debug("unsupport traffic mode: %d\n",
+			 mipi_dsi->traffic_mode);
+		return -EINVAL;
+	}
 	writel(0x0, mipi_dsi->mmio_base + DPI_ENABLE_MULT_PKTS);
 
 	writel(mode->upper_margin, mipi_dsi->mmio_base + DPI_VBP);
-- 
1.7.9.5

